### PR TITLE
Telink b91 tlsr9518adk

### PR DESCRIPTION
**Pull request by yurvyn**
Telink B91 SoC and TLSR9518ADK board support:
- new Telink B91 SoC
- new Telink TLSR9518ADK board
- Drivers: Serial, GPIO, PinMux, Entropy, IEEE802154, PWM, Flash, I2C and SPI

See the board documentation for more details: https://github.com/yurvyn/zephyr/blob/telink_b91_tlsr9518adk/boards/riscv/tlsr9518adk80d/doc/index.rst

hal_telink module request:
https://github.com/zephyrproject-rtos/zephyr/issues/34282

### CODE REVIEW

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts
Comment (galak) [inline boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts:91]: Are the interrupt numbers actually board specific, that seems odd?

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d_defconfig
Comment (galak) [inline boards/riscv/tlsr9518adk80d/tlsr9518adk80d_defconfig:15]: This shouldn't be needed here.

**Review by galak — CHANGES_REQUESTED** [review]
Can we split this PR into a few different PRs.  I think having the SoC + basic board port that boots with UART  and GPIO & PINMUX if needed.  Than have the other drivers as additional PRs.

File: arch/riscv/Kconfig
Comment (galak) [inline arch/riscv/Kconfig:40]: Is there a reason to make this a Kconfig vs just always supporting it?

File: boards/riscv/tlsr9518adk80d/pinmux.c
Comment (galak) [inline boards/riscv/tlsr9518adk80d/pinmux.c:14]: Can we encode these in devicetree.  Take a look at what we do on STM32 or NXP Kinetsis devicetrees for an example.

**Comment by yurvyn** [comment]
> Can we split this PR into a few different PRs. I think having the SoC + basic board port that boots with UART and GPIO & PINMUX if needed. Than have the other drivers as additional PRs.

**Comment by yurvyn** [comment]
> > Can we split this PR into a few different PRs. I think having the SoC + basic board port that boots with UART and GPIO & PINMUX if needed. Than have the other drivers as additional PRs.

Is it necessary to do it?
I hope having it in one PR cold be more effective and safe time.

**Comment by galak** [comment]
> > >
> 
> Is it necessary to do it?
> I hope having it in one PR cold be more effective and safe time.

The problem is the scope of the PR covers a number of driver maintainers and such so getting approval and review from all will take much longer.  Splitting up the PR will get it merged faster and make it much easier to get proper reviews.

File: arch/riscv/Kconfig
Comment (yurvyn) [reply arch/riscv/Kconfig:40]: Yes,
Using RISC-V global pointer is optional and currently there no RISC-V vendors SoC which are using it.
To use global pointer feature the "__global_pointer$" variable should be defined in linker script.

With this solution global pointer usage is disabled by default and it is backward compatible with all other RISC-V SoC vendors.

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d_defconfig
Comment (yurvyn) [reply boards/riscv/tlsr9518adk80d/tlsr9518adk80d_defconfig:15]: I have just checked it, and the solution is not working whiteout "SERIAL", "CONSOLE" and "UART_CONSOLE" configs. 
Other vendors also are enabling these configs in "defconfig" file.

Only UART_CONSOLE_ON_DEV_NAME is optional. I added this config to easy switch between "uart_0" and "uart_1", for debug reasons.

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts
Comment (yurvyn) [reply boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts:91]: Yes, only IRQs 25, 26 and 27 could be assigned for GPIOs.

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d_defconfig
Comment (galak) [reply boards/riscv/tlsr9518adk80d/tlsr9518adk80d_defconfig:15]: I meant `UART_CONSOLE_ON_DEV_NAME`  We normally leave this to come from devicetree and setting of `zephyr,console`

File: arch/riscv/Kconfig
Comment (yurvyn) [reply arch/riscv/Kconfig:40]: However, if we add "__global_pointer$" variable to the generic RISC-V linker script, it will be supported by all RISC-V vendors automatically and in this case we can make it always supporting.
But, I still preferer to have it user configurable because usage of RISC-V global pointer is optional.

Comment (galak) [reply arch/riscv/Kconfig:40]: Would be good to pull this commit into its own PR to allow review by RISC-V collaborators.

File: boards/riscv/tlsr9518adk80d/pinmux.c
Comment (yurvyn) [reply boards/riscv/tlsr9518adk80d/pinmux.c:14]: During the development, I saw that some vendors are using "devicetree" and some vendors are using "pinmux.c" solution. 
I decided to go with "pinmux.c" solution because I preflare to have all pins configuration in one place. 
As I see currently 62 boards are using this solution.

Is it something what must be fixed? Or, it is just a recommendation?

**Comment by yurvyn** [comment]
Ok, than I will close this PR and will follow you recommendation.

File: boards/riscv/tlsr9518adk80d/pinmux.c
Comment (galak) [reply boards/riscv/tlsr9518adk80d/pinmux.c:14]: This should be addressed as the direction we are moving towards is utilizing devicetree for this purpose.

**Pull request closed without merging**
